cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov-node/sriov_node.sh

```diff
@@ -0,0 +1,99 @@
+  [ "${#sriov_pfs[@]}" -eq 0 ] && echo "FATAL: Could not find available sriov PFs on host" >&2 && return 1
+
+  local pf_name
+  local pf_names=()
+  for pf in "${sriov_pfs[@]}"; do
+    pf_name="${pf%%/device/*}"
+    pf_name="${pf_name##*/}"
+    if [ $(echo "${PF_BLACKLIST[@]}" | grep "${pf_name}") ]; then
+      continue
+    fi
+
+    pfs_names+=( $pf_name )
+  done
+
+  echo "${pfs_names[@]}"
+}
+
+function sriov_node::configure_sriov_pfs_and_vfs() {
+  local -r nodes_array=($1)
+  local -r pfs_names_array=($2)
+  local -r pf_count_per_node=$3
+
+  local -r config_vf_script=$(basename "$CONFIGURE_VFS_SCRIPT_PATH")
+  local pfs_to_move=()
+  local pfs_array_offset=0
+  local node_exec
+
+  # 'iplink' learns which network namespaces there are by checking /var/run/netns
+  mkdir -p /var/run/netns
+  for node in "${nodes_array[@]}"; do
+    prepare_node_netns "$node"
+
+    ## Move PF's to node netns
+    # Slice '$pfs_names_array' to have unique silce for each node
+    # with '$pf_count_per_node' PF's names
+    pfs_to_move=( "${pfs_names_array[@]:$pfs_array_offset:$pf_count_per_node}" )
+    echo "Moving '${pfs_to_move[*]}' PF's to '$node' netns"
+    for pf_name in "${pfs_to_move[@]}"; do
+      move_pf_to_node_netns "$node" "$pf_name"
+    done
+    # Increment the offset for next slice
+    pfs_array_offset=$((pfs_array_offset + pf_count_per_node))
+
+    # KIND mounts sysfs as read-only by default, remount as R/W"
+    node_exec="docker exec $node"
+    $node_exec mount -o remount,rw /sys
+    $node_exec chmod 666 /dev/vfio/vfio
+
+    # Create and configure SRIOV Virtual Functions on SRIOV node
+    docker cp "$CONFIGURE_VFS_SCRIPT_PATH" "$node:/"
+    $node_exec bash -c "DRIVER=$VFS_DRIVER DRIVER_KMODULE=$VFS_DRIVER_KMODULE ./$config_vf_script"
+
+    _kubectl label node $node $SRIOV_NODE_LABEL
+  done
+}
+
+function prepare_node_netns() {
+  local -r node_name=$1
+  local -r node_pid=$(docker inspect -f '{{.State.Pid}}' "$node_name")
+
+  # Docker does not create the required symlink for a container netns
+  # it perverts iplink from learning that container netns.
+  # Thus it is necessary to create symlink between the current
+  # worker node (container) netns to /var/run/netns (consumed by iplink)
+  # Now the node container netns named with the node name will be visible.
+  ln -sf "/proc/$node_pid/ns/net" "/var/run/netns/$node_name"
+}
+
+function move_pf_to_node_netns() {
+  local -r node_name=$1
+  local -r pf_name=$2
+
+  # Move PF to node network-namespace
+  ip link set "$pf_name" netns "$node_name"
+  # Ensure current PF is up
+  ip netns exec "$node_name" ip link set up dev "$pf_name"
+  ip netns exec "$node_name" ip link show
+}
+
+function sriov_node::total_vfs_count() {
+  local -r node_name=$1
+  local -r node_pid=$(docker inspect -f '{{.State.Pid}}' "$node_name")
+  local -r pfs_sriov_numvfs=( $(cat /proc/$node_pid/root/sys/class/net/*/device/sriov_numvfs) )
```

Comment:
we have just one node here, so why use an array ?
(also line 94)

just need to make sure it fails once changed, if it doesn't find the sriov_numvfs,
because moving PF to the node is an async command which might take some time to fnish

Reply:
I read it wrong, the array is indeed because there might be theoretical couple PFs per node.

The issue i possibly see here, is since the ip link set is not blocking, and the waiting of it was removed [1]
the amount of VF that it will calculate might bot be accurate, and later when it spin and wait for this number of PF to appear in the resource, it might not be accurate.
For example if there are 2 PFs, it won't exit here with an error, but it will fail the cluster-up with error,
or even worse, it won't fail it with error, because lets say each PF and N VFs,
it will wait for N instead of 2N, and once there are N it will finish successfully, without checking that the other indeed are valid in the resource.

The solution of this can be the following:
Return the code that waits for the PFs to propagate in the node [1] 
but in case it failed the two checks there, return error, unless a flag say PARALLEL=true
and then it can continue as it was
(but in that case, it should also know what is the expected PF it passed, so it can sanity check it indeed catch all their VF in this function - which is good to have anyhow, in all cases)

It will also solve
https://github.com/kubevirt/kubevirtci/pull/570#discussion_r609437642
if we will have PARALLEL=true

[1] https://github.com/kubevirt/kubevirtci/pull/570/files#diff-499cff045d8a875e8b39f4a9fd59354eb99e8f8d91a95ebc70fbf42d2d5b5beaL51

Reply:
> Regarding one PF or more, this function seems not to assume anything, it just returns a number. If that is correct or not, is a question about what is assumed at the caller. If we have two PF/s, is the caller fine to continue or not?

This function calculates how many VF's there are on a node regardless of how many PF's if has.
The [caller](https://github.com/kubevirt/kubevirtci/blob/0645fa84c24db03f09d8ff5fd33dfff41c78d78f/cluster-up/cluster/kind-k8s-sriov-1.17.0/config_sriov.sh#L26)  checks each sriov capable node VF's count against the node allocatable resource that `sriov-device-plugin` updated.

This is how we check that each sriov capable node has the correct allocateable resource
with the correct amount according to sriov-device-plugin config. 
If there will be a problem with missing the PF's on the node or configuring the VF's, 
`sriov-dp` would raise an error and the allocateable resource would not apear on the node state, causing the caller to fail and stop `cluster-up`.